Adding content for `What is serverless?` (#23)
MaximRouiller/AzureFunctions-Workshop: src/content/workshops/functions/1-what-is-serverless/_index.md

outputs: ["Reveal"]
---

## What is Serverless?

- Execution model
- Pay as you consume
- No need to control scaling

{{% note %}}
Serverless is an execution model. Everytime you hear the word serverless, it's talking about this cloud focused execution model. Running the same code locally is a different concept.

An attribute of that execution model is that you pay for what you consume and not for what you reserve. When you create services in cloud providers, you reserve for a certain amount of CPU/RAM. This is not a necessary concern when you start a new serverless application.

Serverless goes without this as the whole premise to care about servers less. Meaning that when using a serverless framework, you should not be required to control scaling at all.

Now, how do you run code on a Serverless platform? This takes us to...

{{% /note %}}

---

## Serverless Runtime

Available Runtimes:

- Azure Functions
- AWS Lambda
- Google Cloud Functions
- IBM Cloud Functions

> - ...

Reviewer: OpenFaas could be interesting to mention as well.

Author: Good catch. Might as well add an open-source project to this as well.

Also called Functions as a Service (**FaaS**)

{{% note %}}

The runtime. The runtime is the code behind that runs your code. Every cloud provider offers their own but what is important to note is that it's different from Serverless. Those runtime may allow you to run your code locally or in an orchestrator.

{{% /note %}}

---

## Azure Functions as a Serverless Runtime

- Azure Functions is the Runtime
- Azure Functions Consumption Plan is Serverless
- Every Azure Functions can run locally, in the Cloud, and on Kubernetes  

> 

Reviewer: One thing I struggle with is when to explain the difference between Azure Functions the service, a Function App, and individual functions and how they work together. Not sure if you want to mention it or not yet. As a complete beginner, it's not needed, but the moment you start working in them, the concepts become important.

Reviewer 2: I would mention that you can host or execute using Consumption Plan or App Service Plan. It worths to describe the difference.

Author: I'll leave it a little bit later. The point here is to explain what is serverless. Once the concept is there, it becomes way easier to understand the difference between Azure Functions, Azure Functions App, and an Azure Functions Function.

{{% /note %}}

---

## Why Serverless?

{{% note %}}

So now that we know what serverless is, why use it? Why not just create a single VM and go from there?

{{% /note %}}
---

## Why Serverless?

### Advantages

- Cost
- No concern about scaling, the cloud takes care of it
- Simplified development model

{{% note %}}
Let's start with the obvious. Cost. Since resources are not reserved, you are only getting invoiced for what you are using and most cloud providers offer you a free quota.

This allows you to build an application without using a single virtual machine. A minimum viable product can be hosted for less than the cost of the milk you put in your coffee.

Then, without having to care about scaling and letting your cloud provider take care of it, you can keep on focusing on your application instead of on the infrastructure.

Finally, the simplified development model means that you work on smaller/simpler units. Meaning that your code is simpler to work, deploy, and scale.

So, to all advantages... we must bring the other side of the coin.
{{% /note %}}

---

## Why Serverless?

### Disadvantages

- Cold-start/Performance
- Simplified development model

> - Potential Vendor Lock-in

Reviewer: Another disadvantage is that it can be difficult to manage a large number of functions, both from a development perspective, as well as operations.

Reviewer 2: I don't see this as a disavatage. It's all about how you structure / package your functions.

Author: Is it the same for all other platforms as well? I'm going for a cloud agnostic definition of Serverless in those slides.

Author: @thdotnet Excellent subject to cover later as well.

{{% note %}}
Among the disadvantages of running serverless workloads includes Cold-Start/Performance.

Running purely in a consumption plan where you can scale all the way down to 0 servers means that your application could be restarting from scratch with nothing in memory. Hence the `cold-start` meaning. Azure solves this partially by allowing an hybrid of consumption and reserved compute which we can talk about later.

> With a simplified development model brings its own issues. Things being simpler means that you can tweak and customize everything as easily as if you had 100% control of the application.

Reviewer: > Things being simpler means that you can tweak and customize everything as easily as if you had 100% control of the application.

I think you meant can't tweak?

Author: Fixed.

outputs: ["Reveal"]
---

## What is Serverless?

- Execution model
- Pay as you consume
- No need to control scaling
- Event Driven

{{% note %}}

> Serverless is an execution model. Everytime you hear the word serverless, it's talking about this cloud focused execution model. Running the same code locally is a different concept.

Reviewer 3: ```suggestion
Serverless is an execution model. When you hear the word "serverless", it's talking about a cloud-focused execution model. Running the same code locally is a different concept.
```
Do you really mean locally here? You can run functions locally for development and testing. I think you more mean running the same logic on your own application server.... or something like that, right?

Author: I mean as during the course of development. Not for production of course. Merging your changes.

> An attribute of that execution model is that you pay for what you consume and not for what you reserve. When you create services in cloud providers, you reserve for a certain amount of CPU/RAM. This is not a necessary concern when you start a new serverless application.

Reviewer 3: ```suggestion
An attribute of the execution model is that you pay for what you consume rather than for reserved resources. When you create services in cloud providers, you reserve for your application a certain amount of CPU/RAM. This is not a necessary concern when you start a new serverless application.
```

When you say, 

> This is not a necessary concern when you start a new serverless application.

Are you trying to make the point that at the beginning of the project most devs aren't thinking about deployment, servers, and resources?

Author: Yes.

> Serverless goes without this as the whole premise to care about servers less. Meaning that when using a serverless framework, you should not be required to control scaling at all.

Reviewer 3: ```suggestion
The point of a serverless application is to focus on your application code, and less about the underlying infrastructure. 
```
Regarding this sentence:

> Meaning that when using a serverless framework, you should not be required to control scaling at all.

There are some relevant scaling concerns that customers may consider. Do you think there may be a better way to phrase this than "you should not be required to control scaling at all"?

Author: I love the way you phrased it.

Finally, everything in serverless is in someway related to events. Whether we are talking about an HTTP request, a timer, a new file is added to a cloud storage, or a message to unqueue, it's all concepts found in serverless. Those events allow the Cloud provider to detect when it needs to provision more or less resources for your application.

Now, how do you run code on a Serverless platform? This takes us to...

{{% /note %}}

---

## Serverless Runtime

Available Runtimes:

- Azure Functions
- AWS Lambda
- Google Cloud Functions
- IBM Cloud Functions
- OpenFaaS
- ...

Also called Functions as a Service (**FaaS**)

{{% note %}}

> The runtime. The runtime is the code behind that runs your code. Every cloud provider offers their own but what is important to note is that it's different from Serverless. Those runtime may allow you to run your code locally or in an orchestrator.

Reviewer 3: This paragraph is confusing to me. Could you re-phrase, perhaps?

Author: Rephrased.

{{% /note %}}

---

## Azure Functions as a Serverless Runtime

- Azure Functions is the Runtime
- Azure Functions Consumption Plan is Serverless
- Every Azure Functions can run locally, in the Cloud, and on Kubernetes  

{{% note %}}

What interests us today is Azure Functions. Functions is the Microsoft FaaS implementation and using the Consumption Plan on Azure is the serverless implementation. Every Function that you write today will run on your local machine as well as in the Cloud.

{{% /note %}}

---

## Why Serverless?

{{% note %}}

So now that we know what serverless is, why use it? Why not just create a single VM and go from there?

{{% /note %}}
---

> ## Why Serverless?

Reviewer 3: I think you can remove this as you have the same title on line 64.

Author: Removed.
